hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/storage/HoodieSparkParquetReader.java

```diff
@@ -86,13 +87,20 @@ public ClosableIterator<InternalRow> getInternalRowIterator(Schema readerSchema,
     conf.setBoolean(SQLConf.PARQUET_BINARY_AS_STRING().key(), (Boolean) SQLConf.get().getConf(SQLConf.PARQUET_BINARY_AS_STRING()));
     conf.setBoolean(SQLConf.PARQUET_INT96_AS_TIMESTAMP().key(), (Boolean) SQLConf.get().getConf(SQLConf.PARQUET_INT96_AS_TIMESTAMP()));
     InputFile inputFile = HadoopInputFile.fromPath(path, conf);
-    ParquetReader reader = new ParquetReader.Builder<InternalRow>(inputFile) {
+    ParquetReader<InternalRow> reader = new ParquetReader.Builder<InternalRow>(inputFile) {
       @Override
       protected ReadSupport getReadSupport() {
         return new ParquetReadSupport();
       }
     }.withConf(conf).build();
-    ParquetReaderIterator<InternalRow> parquetReaderIterator = new ParquetReaderIterator<>(reader, InternalRow::copy);
+    ParquetReaderIterator<InternalRow> parquetReaderIterator = new ParquetReaderIterator<>(reader,
```

Comment:
We should not be making copies in a blanket fashion like that. Why do we need this?

Reply:
1. If parquetReader return UnsafeRow, we should copy it. Because they share the memory. If not copy it, it will fail in UT.
2. If parquetReader return GenericInternalRow, we should unsafe project it and then copy. Because we use the same UnsafeProjection. If not copy it, it will fail in UT.

Reply:
Great idea. I found HoodieMergeHelper#runMerge seems not need InternalRow#copy. Unfortunately, BoundedInMemoryExecutor write the iter in the queue. For the rest, I didn't find the place that need not to copy.

Reply:
@wzx140 that's exactly what i'm referring to -- let's move such copying to the least possible scope to make sure we're only making copies where we absolutely have (and nowhere else)